### DISCUSSION TITLE
I'm finding setting up Vaultwarden with HTTPS very confusing.

### DISCUSSION THREAD

**Discussion by Shadow8472** [dani-garcia/vaultwarden#2140] [Q&A] [answered]
I'm trying to setup Vaultwarden for my second time, upgrading from a Pi 4 setup that broke a while back. I'm trying to set it up "properly" this time without using self-signed certificates this time if reasonably possible. Edit1: (The scope I'm working at is a hobbyist level for home use.)

First off, I know about Let's Encrypt and the guide on how to get it running. [https://github.com/dani-garcia/vaultwarden/wiki/Running-a-private-vaultwarden-instance-with-Let%27s-Encrypt-certs](https://github.com/dani-garcia/vaultwarden/wiki/Running-a-private-vaultwarden-instance-with-Let%27s-Encrypt-certs) It doesn't make sense to me. I've spent months trying to puzzle it out with little success finding literature on HTTPS that doesn't lose me before it communicates its payload to where I can follow the above instructions.

With that said:
- Rocky Linux 8.4 (RHEL 8 compatible)
- LAN only deployment
- Vaultwarden already running in underprivileged Podman container with a dedicated, locked user (I got the HTTP login page)
- Host will need ports 80 and 443 for other projects (plus non-root container).

So far, I estimate I understand about half the process of getting a Let's Encrypt certificate on a conceptual level. I do own my own domain with a paid certificate through a hosting service, so I won't need to register at DuckDNS or Cloudflare.

The only reason I've kept with trying to do HTTPS properly is because "it's the right thing to do" or some variant thereof. Why should self-signing only be used for testing besides the "hassle" of installing the certificate across half a score of devices? What about using Diffie-Hellman Key Exchange as described here: [https://bitwarden.com/help/article/certificates#using-diffie-hellman-key-exchange](https://bitwarden.com/help/article/certificates#using-diffie-hellman-key-exchange)? Is that even implemented?

**Comment by RealOrangeOne** [comment, marked as answer]
Diffie Hellman is only useful if you are having Bitwarden generate the certificates for you, which you aren't. 

If you need port 80 and 443 reserved for other projects, then just use a different port.

I think the issue is that it's not as simple as "just having HTTPS". To have a full certificate for a LAN-only application, you also need a custom DNS server to override the domain you want to use. Certificates have to have a domain, so you need to have a domain which will resolve to your server (hence the custom DNS server). Once that's working, you can get a certificate, which will need to be with letsencrypts DNS challenge rather than HTTP as you won't be internet accessible. 

For the vaultwarden-specific bits, you'll want to [generate a certificate using `lego`](https://github.com/dani-garcia/vaultwarden/wiki/Running-a-private-vaultwarden-instance-with-Let%27s-Encrypt-certs#getting-certs-using-the-lego-cli), and [pass those certificates to vaultwarden](https://github.com/dani-garcia/vaultwarden/wiki/Enabling-HTTPS#via-rocket).

Besides those 2 above links, the rest is completely separate to vaultwarden, and in fact is exactly the same no matter what application you were trying to use.

**Reply by Shadow8472** [reply]
Thanks for getting back.

I forgot to mention that this is for home use. I don't have a domain (that I know of, I haven't researched those yet) and setting one up special is a little out of my scope for now. I'll also point out that I'm not committed to any one method at this time. If Diffie Hellman will be easier to figure out than extracting my blog's certificate (and setting up other required network infrastructure), that's what I'll do.

The most important question I have is: why should I even bother with "doing it right" at a hobby level? What is so wrong with using a self-signed certificate over a LAN (my home network) that one should only be used for testing?

**Reply by BlackDex** [reply]
Nothing wrong with self-signed for home or testing, only issues is that since some clients do not support or at least not easy these kind of certificates which may degrade your user experience too.

If you do not need Webauthn/Fido and only use a browser, then go ahead and follow the self signed steps.

Only thing i do advice is to use a reverse proxy like nginx or caddy or haproxy. Whatever you prefer to use and configure the ssl stuff there.

Regarding the dhparam. You are mix and matching stuff which shouldn't. That article is meant for Bitwarden and not Vaultwarden. The dhparam is used for nginx, which is used by the Bitwarden project. It can be used for Vaultwarden, but for that you need to use nginx as a reverse proxy.

**Reply by Shadow8472** [reply]
If I understand everything correctly:
- A self-signed certificate will be just fine for browser clients, but standalone clients and 2FA systems may throw a fit.
- I can configure a reverse proxy to vouch for a self-signed certificate.
- This reverse proxy can sit on the same host as Vaultwarden, listening to opened ports facing the wider home network and directing Vaultwarden traffic to an internal address.
- Vaultwarden does have Diffie-Hellman Key Exchange (dhparam) implemented but among the listed reverse proxies listed, only nginx is compatible.

**Reply by BlackDex** [reply]
> If I understand everything correctly:
> 
> * A self-signed certificate will be just fine for browser clients, but standalone clients and 2FA systems may throw a fit.

Correct

> * I can configure a reverse proxy to vouch for a self-signed certificate.

Not sure what you mean with vouch here. The reverse proxy should handle the SSL connection and then passes out unsecured to Vaultwarden.

> * This reverse proxy can sit on the same host as Vaultwarden, listening to opened ports facing the wider home network and directing Vaultwarden traffic to an internal address.

Correct 

> * Vaultwarden does have Diffie-Hellman Key Exchange (dhparam) implemented but among the listed reverse proxies listed, only nginx is compatible.

Wrong. Every SSL supporting application has DH support. It can be in many bit depths. By default nginx uses 1024, which is a bit low. Other applications or reverse proxies use a higher depth by default and actually do not provide a way to increase it, like caddy for example. And the same goes for the library used by Vaultwarden (or Rocket actually, rustls). It doesn't even support that at all because the author find those insecure.

**Reply by Shadow8472** [reply]
2 out of 4 isn't all that bad, I suppose.

> Not sure what you mean with vouch here. The reverse proxy should handle the SSL connection and then passes out unsecured to Vaultwarden.
Ugg! Words. I feel like a picture would be helpful here. Before I was figuring something like the following:
- Bitwarden client connects to reverse proxy
- Reverse proxy forwards connection to Vaultwarden's container
- Vaultwarden holds SSL certificate
- Reverse proxy does some stuff so Bitwarden client accepts secure connection


Now, my understanding is:
- Bitwarden client connects to reverse proxy
- Reverse proxy hosts SSL certificate (or otherwise contacts a separate service also on the host)
- Reverse proxy (or the client?) contacts an authority outside the local network. (Can you have HTTPS on a network separated by an air gap At the moment, I don't think so.)
- Reverse proxy forwards Bitwarden client's traffic to Vaultwarden's container without SSL

I might be forgetting a thing or two, but time crunch. Thanks again for your help. I'm thinking the other point I made a guess on may be a bit above my understanding for now.

**Reply by BlackDex** [reply]
It's rather simple actually. And the reason we suggest to use a reverse proxy is that the build-in SSL library (at least in the current latest version of Vaultwarden) could have some issue with some clients because it doesn't support all protocols. Therefore a reverse proxy application is build just for that and is able to provide better security for these SSL connections.

Regarding the flow.
<pre>
Client < SSL/HTTPS > Reverse-Proxy < Plain/HTTP > Valtwarden
</pre>
So, the client connects with a secure connection to the reverse proxy, the reverse proxy connects using simple HTTP to Vaultwarden.
The reverse proxy handles all the SSL encryption.

It doesn't matter if that flow is in an air-gapped environment or out in the open, or just in your home network with internet connection, but no incoming opened ports.

The only thing which is not possible within a fully air-gapped environment is for a real valid certificate to check its OCSP stapling, to verify if the certificate is still valid, but for a self-signed cert that is never going to happen anyway unless you run your own OCSP server, but that would be too much.

**Reply by Shadow8472** [reply]
Thanks a bundle. I still don't understand the full process, but that will probably need to wait until I inevitably revisit this or another project again. I'm marking this thread as answered.
